### PR TITLE
KFold cross-validation

### PR DESCRIPTION
**Pull request by Mec-iS**
#5 

...
* implement `test_indices`, `test_matrices`, `split` to accept a matrix and return a `Vec`
* improve allocation (?)
* make all three to return an iterator (?)

### CODE REVIEW

File: src/model_selection/mod.rs [hunk 1]
```diff
@@ -107,3 +107,43 @@ mod tests {
         assert_eq!(x_test.shape().0, y_test.len());
     }
 }
+
+
+///
+/// KFold Cross-Validation
+///
+/// Entities involved in the KFold procedure:
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:115]: Will our user benefit from knowing the list of entities in the KFold procedure? 

Doc strings that start from `///` will be displayed in https://docs.rs/smartcore/0.1.0/smartcore/model_selection/index.html and it is important to keep description of the methods and classes short but informative. The danger of keeping a lot of unnecessary information here is that this information might get out-of-sync if not properly maintained. Doc strings should describe the function of the method or a class, meaning of all parameters, some technical details like errors and default values and a short example. Any other information that describes algorithm in greater details should go into the user's manual: https://github.com/smartcorelib/smartcorelib.org

File: src/model_selection/mod.rs [hunk 2]
```diff
@@ -107,3 +107,43 @@ mod tests {
...
+///     * a dataset
+///     * a number k of groups (k-folds)
+/// 
+/// Procedure in `cross_validate()`:
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:119]: user's manual: https://github.com/smartcorelib/smartcorelib.org is a better place for the algorithm's pseudocode.

**Review by VolodymyrOrlov — CHANGES_REQUESTED** [review]
Hi @Mec-iS, this is a great start! I have couple of suggestions for you that will help to improve this PR:

* Run `cargo fmt` every time before committing your code, otherwise automatic build will fail if code is incorrectly formatted. 
* Test how project's documentation looks by running `cargo doc --no-deps --features "nalgebra-bindings ndarray-bindings"`. This command will generate html with documentation in the `target/doc/` folder. 

I am looking forward to see implementation of the function that splits dataset into k folds!

**Comment by Mec-iS** [comment]
Thanks. As a next step I will try to implement the interface with only the matrix as input (not considering shuffle and groups).

**Comment by Mec-iS** [comment]
@VolodymyrOrlov can you specify the meaning of [this assignment](https://github.com/scikit-learn/scikit-learn/blob/0864c5896804bc1066c02bcb1443c962cabdc420/sklearn/model_selection/_split.py#L439)

**Comment by VolodymyrOrlov** [comment]
I think the idea is to make sure that total number of samples from all folds equals total number of samples. Here is an example
image
without this line we have only 2 + 2 + 2 = 6 samples in all folds. If you uncomment this line and run the code again total number of samples equal number of samples in X, 3 + 2 + 2 = 7:
image

**Comment by Mec-iS** [comment]
Please check that `test_indices` works as expected

**Comment by Mec-iS** [comment]
Ready for review :+1:

File: src/model_selection/mod.rs [hunk 3]
```diff
@@ -81,6 +81,100 @@ pub fn train_test_split<T: RealNumber, M: Matrix<T>>(
     (x_train, x_test, y_train, y_test)
 }
 
+///
+/// KFold Cross-Validation
+///
+trait BaseKFold {
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:87]: \+ pub, otherwise the trait `BaseKFold` is invisible outside of this module
```suggestion
pub trait BaseKFold {
```

File: src/model_selection/mod.rs [hunk 4]
```diff
@@ -81,6 +81,100 @@ pub fn train_test_split<T: RealNumber, M: Matrix<T>>(
...
+    /// Return a tuple containing the the training set indices for that split and
+    /// the testing set indices for that split.
+    fn split<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<(Vec<usize>, Vec<usize>)>;
+}
+
+///
+/// An implementation of KFold
+///
+pub struct KFold {
+    n_splits: i32, // cannot exceed std::usize::MAX
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:103]: \+ pub, unless you want to add a public method `KFold::new` that will create an instance of this class

```suggestion
    pub n_splits: i32, // cannot exceed std::usize::MAX
```

File: src/model_selection/mod.rs [hunk 5]
```diff
@@ -81,6 +81,100 @@ pub fn train_test_split<T: RealNumber, M: Matrix<T>>(
...
+            // set mask's indices to true according to test indices
+            for i in test_index {
+                test_mask[i] = true; // can be implemented with map()
+            }
+            return_values.push(test_mask);
+        }
+        return return_values;
+    }
+
+    fn split<T: RealNumber, M: Matrix<T>>(&self, x: &M) -> Vec<(Vec<usize>, Vec<usize>)> {
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:151]: Looks like this method does not split **x** in the same way as Scikit  does. For example, 3-fold split of 3x4 matrix in Python would look like this:
image. 

While our implementation outputs this:
image
image

I suggest this test to verify parity between Scikit learn implementation and ours:
```suggestion
#[test]
    fn numpy_parity_test() {
        let k = KFold { n_splits: 3 };
        let x: DenseMatrix<f64> = DenseMatrix::rand(10, 4);
        let expected: Vec<(Vec<usize>, Vec<usize>)> = vec!(
            (vec!(4, 5, 6, 7, 8, 9), vec!(0, 1, 2, 3)),
            (vec!(0, 1, 2, 3, 7, 8, 9), vec!(4, 5, 6)),
            (vec!(0, 1, 2, 3, 4, 5, 6), vec!(7, 8, 9))
        );
        for ((train, test), (expected_train, expected_test)) in k.split(&x).into_iter().zip(expected) {
            assert_eq!(test, expected_test);
            assert_eq!(train, expected_train);
        }        
    }
```

**Review by VolodymyrOrlov — CHANGES_REQUESTED** [review]

**Comment by VolodymyrOrlov** [comment]
@Mec-iS the code looks good! Could of things left before I can proceed to merging this PR:

1. Looks like our implementation does not split data the same way as Scikit-learn's does. 
1. Arguably flag `shuffle` is the most important feature of K-fold split. Let's add it and set it to true by default.

**Comment by Mec-iS** [comment]
I had problem understanding the shape of the matrix, in my understanding:
* `M::shape().0` returns the dimensions (columns)
* `M::shape().1` returns the samples (rows)
Please provide guidance.

**Comment by Mec-iS** [comment]
ahah, sorry I got the dimensions confused. I just had to switch which index in `shapes` I was using and now the tests seem to pass.

File: src/model_selection/mod.rs (hunk 4 above)
Comment (Mec-iS) [reply src/model_selection/mod.rs:103]: I don't see any usage pattern that requires `n_splits` to be accessible outside of the type.

**Comment by Mec-iS** [comment]
Question: in my [last commit I added this line](https://github.com/smartcorelib/smartcore/pull/8/files#diff-ca3045cade7b17a743956d16a5d135deR153) to explicitly define `return_values` as immutable before returning ownership, the same for `indices` in `test_indices` after the optional shuffle; is this worth it or the compiler can do this kind of optimisation automatically? I couldn't find a reference to this in Rust doc so I am not sure. If not, let me know I will revert the commit.

**Comment by VolodymyrOrlov** [comment]
> I had problem understanding the shape of the matrix, in my understanding:
> 
> * `M::shape().0` returns the ~dimensions (columns)~ the samples
> * `M::shape().1` returns the ~samples (rows)~ the dimensions
> 
> Please provide guidance. I couldn't find this information clearly stated in the documentation.

Good catch, thanks for pointing it out! I'll fix documentation of the `shape` method

**Comment by VolodymyrOrlov** [comment]
> Question: in my [last commit I added this line](https://github.com/smartcorelib/smartcore/pull/8/files#diff-ca3045cade7b17a743956d16a5d135deR153) to explicitly define `return_values` as immutable before returning ownership, the same for `indices` in `test_indices` after the optional shuffle; is this worth it or the compiler can do this kind of optimisation automatically? I couldn't find a reference to this in Rust doc so I am not sure. If not, let me know I will revert the commit.

No, it is not necessary since you pass ownership of the `return_values` to the caller. A function that calls the method `test_indices` will be able to modify `return_values` and I see no reason why we should not let it to do so. 

I am sure you've read this book but will leave this link here in case you haven't: https://doc.rust-lang.org/book/ch04-00-understanding-ownership.html.

File: src/model_selection/mod.rs [hunk 6]
```diff
@@ -81,6 +83,118 @@ pub fn train_test_split<T: RealNumber, M: Matrix<T>>(
...
+                .collect::<Vec<usize>>(); // filter train indices out according to mask
+            let test_index = indices
+                .iter()
+                .enumerate()
+                .filter(|&(idx, _)| test_index[idx] == true)
+                .map(|(idx, _)| idx)
+                .collect::<Vec<usize>>(); // filter tests indices out according to mask
+            return_values.push((train_index, test_index))
+        }
+        return return_values;
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:194]: nit: you can simplify this to
```suggestion
        return_values
```

File: src/model_selection/mod.rs [hunk 7]
```diff
@@ -81,6 +83,118 @@ pub fn train_test_split<T: RealNumber, M: Matrix<T>>(
...
+        let mut return_values: Vec<Vec<usize>> = Vec::with_capacity(self.n_splits as usize);
+        let mut current: usize = 0;
+        for fold_size in fold_sizes.drain(..) {
+            let stop = current + fold_size;
+            return_values.push(indices[current..stop].to_vec());
+            current = stop
+        }
+
+        let return_values = return_values;
+        return return_values;
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:154]: nit: you can simplify this to
```suggestion
        return_values
```

File: src/model_selection/mod.rs [hunk 8]
```diff
@@ -81,6 +83,118 @@ pub fn train_test_split<T: RealNumber, M: Matrix<T>>(
...
+        for test_index in self.test_indices(x).drain(..) {
+            // init mask
+            let mut test_mask = vec![false; x.shape().0];
+            // set mask's indices to true according to test indices
+            for i in test_index {
+                test_mask[i] = true; // can be implemented with map()
+            }
+            return_values.push(test_mask);
+        }
+        return return_values;
```
Comment (VolodymyrOrlov) [inline src/model_selection/mod.rs:168]: nit: you can simplify this to
```suggestion
        return_values
```

**Comment by Mec-iS** [comment]
> since you pass ownership of the return_values to the caller. A function that calls the method test_indices will be able to modify return_values and I see no reason why we should not let it to do so.

yes but I dont see the necessity of passing the ownership of a mutable vector. It could be an interface leak to allow mutability on something that is not meant to be modified. Once the indices and the splits are defined there is no reason for the developer-user to be able to modify it. That is why is conceptually more correct to return an immutable maybe.

**Comment by VolodymyrOrlov** [comment]
> > since you pass ownership of the return_values to the caller. A function that calls the method test_indices will be able to modify return_values and I see no reason why we should not let it to do so.
> 
> yes but I dont see the necessity of passing the ownership of a mutable vector. It could be an interface leak to allow mutability on something that is not meant to be modified. Once the indices and the splits are defined, there is no reason for the developer-user to be able to modify it. That is why is conceptually more correct to return an immutable maybe.

I agree with you conceptually but I don't know an easy way to do it without wrapping your return value into another struct that will prevent a caller from modifying it. https://users.rust-lang.org/t/how-to-return-an-immutable-reference-of-a-mutable-vector/30773/5

I think wrapping return value into protective shell is an overkill

**Comment by Mec-iS** [comment]
>I agree with you conceptually but I don't know an easy way to do it without wrapping your return value into another struct that will prevent a caller from modifying it. https://users.rust-lang.org/t/how-to-return-an-immutable-reference-of-a-mutable-vector/30773/5
I think wrapping return value into protective shell is an overkill

no wait, something missing. I have already implemented it with re-assigning the return value to an immutable variable: `let return_values = return_values`, the right side is a mutable vector but once re-assigned to an immutable variable it gets immutable and later returned by the next line (`return_values`). According to my current knowledge of Rust it should work, I just wanted comfirmation that this makes sense (the compiler does not complain but I didnt check what happens in memory)

**Review by VolodymyrOrlov — APPROVED** [review]
lgtm
